### PR TITLE
feat(automation): auto-clear-blocking-labels.yml workflow (#195)

### LINKED ISSUE #195: [#191 sub-A] auto-clear-blocking-labels.yml workflow
Sub-issue of #191 (Auto-remove `needs-external-review` label by CI when external review clears).

## Scope

Ship the core deliverable: a new GitHub Actions workflow `.github/workflows/auto-clear-blocking-labels.yml` that re-evaluates the merge gate on event triggers and removes `needs-external-review` from a PR when the gate passes. This is the load-bearing piece of #191.

Depends on:
- #194 (resolve-pr-threads.sh fix) — not strictly blocking, but the auto-clear workflow re-uses the same conversation-resolution discoverability assumptions.
- Branch protection's `External Review Check` no longer being a required check (already done in this session — the canonical required-check list is now `["Label Gate", "Self-Review Required", "lint"]`).

### 1. Workflow file

`.github/workflows/auto-clear-blocking-labels.yml`:

```yaml
name: Auto-clear blocking labels

on:
  pull_request_target:
    types: [synchronize, labeled, unlabeled]
  pull_request_review:
    types: [submitted]
  check_suite:
    types: [completed]

permissions:
  pull-requests: write
  contents: read

jobs:
  evaluate:
    name: Re-evaluate codex-review-check gate
    if: |
      github.event_name == 'pull_request_target' ||
      github.event_name == 'pull_request_review' ||
      (github.event_name == 'check_suite' && github.event.check_suite.conclusion == 'success')
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd  # v6.0.2
      - name: Resolve PR number
        id: pr
        run: |
          # Different events surface PR# differently; this normalizes.
          # ...
      - name: Run codex-review-check.sh
        id: gate
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          # Read-only; exits 0 when (a) CI green + (b) reviewer-identity
          # APPROVED OR same-agent + Codex 👍 + (c) Codex cleared on HEAD
          if scripts/codex-review-check.sh "${{ steps.pr.outputs.number }}"; then
            echo "passed=true" >> "$GITHUB_OUTPUT"
          fi
      - name: Remove needs-external-review on pass
        if: steps.gate.outputs.passed == 'true'
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          PR=${{ steps.pr.outputs.number }}
          HAS_LABEL=$(gh pr view "$PR" --json labels --jq '[.labels[].name] | contains(["needs-external-review"])')
          if [ "$HAS_LABEL" = "true" ]; then
            gh pr edit "$PR" --remove-label needs-external-review
            gh pr comment "$PR" --body "Automated label removal: \`codex-review-check.sh\` cleared the merge gate on this HEAD. See [auto-clear-blocking-labels.yml](.github/workflows/auto-clear-blocking-labels.yml) for the trigger doctrine."
          fi
```

### 2. PR-number resolution helper step

The three events surface PR number differently:
- `pull_request_target`: `github.event.pull_request.number`
- `pull_request_review`: `github.event.pull_request.number`
- `check_suite.completed`: NOT a PR event; need to enumerate `github.event.check_suite.pull_requests[]` and re-fire per PR

A single script step `scripts/workflow/resolve-pr-from-event.sh` (or inline jq) handles the dispatch.

### 3. Avoid the `pull_request_target` self-fire loop

`pull_request_target.unlabeled` triggers when this workflow itself removes the label. The job's `if:` must filter that out — easiest check: skip when `github.event.action == 'unlabeled' && github.event.label.name == 'needs-external-review'`. The label-remove action that fires the workflow is the one whose effect we just executed; re-running adds nothing.

### 4. Token scope

`secrets.GITHUB_TOKEN` with `pull-requests: write` is sufficient for `gh pr edit --remove-label`. No reviewer PAT needed. The byline on the label removal is `github-actions[bot]`, which is the correct audit-trail signal ("CI removed this, not a human or agent").

### 5. Test coverage

Add a smoke test in `scripts/ci/check_workflow_events` (new) that validates:
- The PR-number resolution dispatch handles all 3 event shapes
- The self-fire loop guard activates on the synthesized unlabeled-by-self event

Live integration testing happens on the next over-threshold PR after merge.

## Acceptance

- [ ] `.github/workflows/auto-clear-blocking-labels.yml` ships and triggers on `synchronize`, `pull_request_review.submitted`, `check_suite.completed`
- [ ] Workflow successfully removes `needs-external-review` after `codex-review-check.sh` exits 0 on a real PR
- [ ] Self-fire loop guard prevents recursion on `unlabeled` events
- [ ] PR comment is posted on label removal with the workflow attribution
- [ ] (Optional) `scripts/ci/check_workflow_events` smoke test added

## Out of scope

- Doctrine carve-out for the agent-prohibition rule — see #191 sub-B.
- Periodic scheduled re-eval for the 👍-after-last-push case — see #191 sub-C.
- `policy-violation` auto-removal — file as a sibling issue when the policy-check workflow that applies it is identified.

Authoring-Agent: claude

### PR DESCRIPTION
**Pull request by nathanjohnpayne**
Closes #195. Sub-A of parent #191.

## Why

Eliminates the manual `gh pr edit --remove-label needs-external-review` toil that hits every over-threshold PR after external review clears — the friction we just lived through on PR #194 in the same session that filed #195.

## What

New workflow `.github/workflows/auto-clear-blocking-labels.yml`:

- **Triggers**: `pull_request_target` (synchronize, labeled, unlabeled) + `pull_request_review` (submitted) + `check_suite` (completed). The three surfaces that signal "merge gate state may have just changed."
- **Per-PR action**: if `needs-external-review` label present, run `scripts/codex-review-check.sh` (read-only, exits 0 on pass). On exit 0, remove the label via `secrets.GITHUB_TOKEN` and post an attribution comment.
- **Self-fire loop guard**: `if:` skips when the `unlabeled` event is itself for `needs-external-review` (we just removed it).
- **Scope discipline**: ONLY removes `needs-external-review`. `needs-human-review` and `policy-violation` remain manual-only per REVIEW_POLICY.md § Agent prohibitions. The doctrine carve-out for this workflow ships in [#196](https://github.com/nathanjohnpayne/mergepath/issues/196).

## Test coverage

New `scripts/ci/check_auto_clear_workflow` with 10 tests:

- **Structural** (4): triggers present (×3), self-fire guard present, scope discipline (only target label is removed).
- **Dispatch** (5): PR-number resolution for `pull_request_target`, `pull_request_review`, `check_suite` (multi-PR), `check_suite` (empty), unsupported event.
- **Bonus**: optional YAML parse via yq → PyYAML → graceful skip.

Wired into `repo_lint.yml` alongside the other check_* suites.

## Live verification

Happens on the next over-threshold PR after this lands. Expected: human does NOT need to remove `needs-external-review` for the PR to merge.

## Self-Review

- **Correctness**: workflow follows #195's implementation plan exactly. Self-fire guard prevents recursion. Scope-limited to one label per doctrinal constraint.
- **Regression risk**: low — additive new workflow; no changes to existing review/merge surfaces. Worst case if buggy: label stays on PR (current state), human removes it manually (current path).
- **Coverage**: 10 unit tests; wired into CI.
- **Style**: matches existing workflow file shape (pinned action SHA, permissions block, env-driven inputs).
- **Security/dependency hygiene**: uses `secrets.GITHUB_TOKEN` with `pull-requests:write` scope only — minimal permissions per least-privilege.

Authoring-Agent: claude

### CODE REVIEW

File: .github/workflows/auto-clear-blocking-labels.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+# REVIEW_POLICY.md § Agent prohibitions). The doctrine carve-out for
+# this workflow ships in #196.
+
+on:
+  pull_request_target:
+    types: [synchronize, labeled, unlabeled]
+  pull_request_review:
+    types: [submitted]
+  check_suite:
+    types: [completed]
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply .github/workflows/auto-clear-blocking-labels.yml]: Addressed in 01bbea2 — replaced check_suite with workflow_run targeting the three workflows whose required-check signals codex-review-check.sh inspects (PR Review Policy, Agent Review Pipeline, repo-lint). Added a conclusion=success guard so failed CI runs don't spin the gate re-evaluator pointlessly. The 'reviewer approves first → Actions CI catches up' path now produces a workflow_run event when CI completes, firing the auto-clear correctly. Updated test 2 with both positive (workflow_run present) and negative (check_suite absent) assertions to close the regression.

File: .github/workflows/auto-clear-blocking-labels.yml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+              echo "Label needs-external-review not present; nothing to clear."
+              echo "::endgroup::"
+              continue
+            fi
+
+            set +e
+            scripts/codex-review-check.sh "$PR"
+            rc=$?
+            set -e
+
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply .github/workflows/auto-clear-blocking-labels.yml]: Addressed in 01bbea2 — switched to `set -euo pipefail` from the top of the run block (no asymmetric on/off) and wrapped each gh call (view/edit/comment) in explicit error handling: gh pr view failure → warning + continue (skip this PR); gh pr edit failure → warning + continue (don't try to comment if edit failed); gh pr comment failure → warning only (label removal already succeeded, so the comment is best-effort). Transient gh failures on one PR no longer kill the sweep across multiple PRs.

File: scripts/ci/check_auto_clear_workflow [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      ;;
+    check_suite)
+      jq -r '.check_suite.pull_requests[].number' <<<"$event_json"
+      ;;
+    *)
+      echo "Unsupported event: $event_name" >&2
+      return 1
+      ;;
+  esac
+}
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply scripts/ci/check_auto_clear_workflow:242]: Addressed in 01bbea2 — fixed the dispatch stub to return 0 on unsupported events (matches the workflow's `exit 0  # don't fail; just skip` behavior). Updated the matching test assertion. Also added 3 new tests for the workflow_run paths (success+multi-PR, failure-conclusion-skipped, empty-pull_requests) that the trigger swap introduces.

File: .github/workflows/auto-clear-blocking-labels.yml [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        github.event.label.name == 'needs-external-review'
+      )
+    steps:
+      - name: Checkout repo (for codex-review-check.sh)
+        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd  # v6.0.2
+        with:
+          # Need full history so codex-review-check can inspect the
+          # PR HEAD against base. The script doesn't currently do that
+          # but a future tightening might.
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply .github/workflows/auto-clear-blocking-labels.yml:73]: Critical security fix in 721fbcf — pinned actions/checkout to `ref: \${{ github.event.repository.default_branch }}` so the workflow ALWAYS runs the trusted base-branch copy of scripts/codex-review-check.sh, never the PR's version. Added regression test 'checkout pinned to default_branch' that grep-asserts the explicit ref is present. Thanks for catching this — the combination of pull_request_review default ref + this workflow's pull-requests:write scope was a real foothold.

File: .github/workflows/auto-clear-blocking-labels.yml [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                ;;
+              1)
+                echo "Gate not yet met — leaving label in place."
+                ;;
+              3)
+                echo "::error::codex-review-check.sh config/arg error (rc=3) on PR #$PR"
+                # Don't exit non-zero; one bad PR shouldn't kill the
+                # whole sweep on workflow_run events that may have
+                # multiple PRs attached.
+                ;;
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply .github/workflows/auto-clear-blocking-labels.yml:186]: Addressed in 721fbcf — added `had_infra_error=0` sentinel before the loop, set it to 1 in the rc=3 case (continuing past so multi-PR sweeps still cover other PRs), and `exit \"\$had_infra_error\"` after the loop. Workflow now surfaces gate-evaluation infrastructure errors as job failures. New regression test 'rc=3 sets infra-error sentinel + job exits non-zero on it' guards both the sentinel set and the final exit.

File: scripts/ci/check_auto_clear_workflow [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+# doesn't fire for Actions-created suites).
+for trig in 'pull_request_target' 'pull_request_review' 'workflow_run'; do
+  if grep -qE "^[[:space:]]+${trig}:" "$WORKFLOW"; then
+    pass=$((pass + 1))
+    echo "  PASS: trigger '$trig' configured"
+  else
+    fail=$((fail + 1))
+    echo "  FAIL: missing required trigger '$trig'" >&2
+  fi
+done
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply scripts/ci/check_auto_clear_workflow:60]: Addressed in 721fbcf — added per-workflow-name assertions for 'PR Review Policy', 'Agent Review Pipeline', and 'repo-lint'. Typos or renames on either side now fail the smoke test. 17/17 tests pass.

File: .github/workflows/auto-clear-blocking-labels.yml [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+              3)
+                echo "::error::codex-review-check.sh config/arg error (rc=3) on PR #$PR"
+                # Don't exit non-zero mid-loop — keep sweeping other
+                # PRs on a multi-PR workflow_run trigger. But set
+                # the sentinel so the job exits non-zero at the end.
+                had_infra_error=1
+                ;;
+              *)
+                echo "::warning::codex-review-check.sh unexpected rc=$rc on PR #$PR"
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply .github/workflows/auto-clear-blocking-labels.yml]: Addressed in 961dc9b — `*)` case now sets `had_infra_error=1` (same sentinel as rc=3). Any rc outside 0/1/3 surfaces as job failure. Better to fail loudly than leave the label stuck behind a silently-broken gate.

File: scripts/ci/check_auto_clear_workflow [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+# infrastructure failures would silently leave the workflow green.
+# CR Major on PR #202 r2.
+if grep -qF "had_infra_error=1" "$WORKFLOW" && \
+   grep -qE 'exit "?\$had_infra_error"?' "$WORKFLOW"; then
+  pass=$((pass + 1))
+  echo "  PASS: rc=3 sets infra-error sentinel + job exits non-zero on it"
+else
+  fail=$((fail + 1))
+  echo "  FAIL: rc=3 from codex-review-check.sh must surface as job failure (sentinel + final exit)" >&2
+fi
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply scripts/ci/check_auto_clear_workflow:135]: Addressed in 961dc9b — added a least-privilege test that requires `pull-requests:write` + `contents:read` + `checks:read` AND rejects any other `:write` permission across the documented enumeration (actions, attestations, deployments, discussions, id-token, issues, models, packages, pages, security-events, statuses). Future scope expansion now surfaces as a CI failure.

File: scripts/ci/check_auto_clear_workflow [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+  echo "  PASS: self-fire loop guard for unlabeled+needs-external-review present"
+else
+  fail=$((fail + 1))
+  echo "  FAIL: self-fire loop guard missing or malformed" >&2
+fi
+
+# ── Test 4: scope discipline — only needs-external-review is removed
+if grep -qE 'remove-label needs-external-review' "$WORKFLOW" && \
+   ! grep -qE 'remove-label needs-human-review' "$WORKFLOW" && \
+   ! grep -qE 'remove-label policy-violation' "$WORKFLOW"; then
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply scripts/ci/check_auto_clear_workflow]: Addressed in 961dc9b — tightened both positive and negative greps from `remove-label <label>` to `gh pr edit .+--remove-label <label>`. Anchors on the executable line; comment/attribution text alone no longer satisfies the assertion.

File: scripts/ci/check_auto_clear_workflow [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+# Least-privilege permissions: only the expected scopes should be
+# present. The security pin above guards checkout but doesn't catch
+# the case where the workflow later gains broader write scopes
+# (e.g., `actions: write`, `issues: write`, etc.). CR Major on PR
+# #202 r3 — codify the expected permission set so any future
+# scope expansion is surfaced as a CI failure.
+if grep -qE '^[[:space:]]+pull-requests:[[:space:]]+write$' "$WORKFLOW" && \
+   grep -qE '^[[:space:]]+contents:[[:space:]]+read$' "$WORKFLOW" && \
+   grep -qE '^[[:space:]]+checks:[[:space:]]+read$' "$WORKFLOW" && \
+   ! grep -qE '^[[:space:]]+(actions|attestations|deployments|discussions|id-token|issues|models|packages|pages|security-events|statuses):[[:space:]]+write$' "$WORKFLOW"; then
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply scripts/ci/check_auto_clear_workflow]: Addressed in 98e703d — switched from a denylist to an exact-set comparison via awk parse of the permissions block. Reads everything between `permissions:` and the next top-level key, sorts alphabetically, string-equality against the expected set:

```
checks: read
contents: read
pull-requests: write
```

Any drift (added scope, changed read↔write, removed scope) fails with a clear expected-vs-got diff. The denylist's gap on artifact-metadata / repository-projects / vulnerability-alerts is closed by definition.

File: scripts/ci/check_auto_clear_workflow [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+  echo "      expected:" >&2; echo "$expected_perms" | sed 's/^/        /' >&2
+  echo "      got:" >&2; echo "$perm_block" | sed 's/^/        /' >&2
+fi
+
+# Sentinel: rc=3 from codex-review-check.sh sets had_infra_error and
+# the job exits non-zero at the end. Without this, gate-evaluation
+# infrastructure failures would silently leave the workflow green.
+# CR Major on PR #202 r2.
+if grep -qF "had_infra_error=1" "$WORKFLOW" && \
+   grep -qE 'exit "?\$had_infra_error"?' "$WORKFLOW"; then
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply scripts/ci/check_auto_clear_workflow]: Addressed in 32ea1a4 — same pattern as the r3 scope-discipline fix. Both greps anchored to `^[[:space:]]*<token>([[:space:]]|$)` so only executable lines satisfy. Comment text or commented-out occurrences won't pass anymore.

**Review by nathanpayne-cursor — APPROVED** [review]
Approved as nathanpayne-cursor on HEAD 8a26f4e6660090fdaa3c2e19acd34c758521ea3a.

Reviewed the auto-clear workflow and the test coverage after the CodeRabbit/Codex rounds. The current head pins checkout to the trusted default branch, listens on `workflow_run` rather than `check_suite`, keeps the label scope limited to `needs-external-review`, fails closed on gate-script infrastructure/contract errors, validates permissions as an exact set, and includes bash syntax coverage for the embedded run blocks.

Local verification:
- `./scripts/ci/check_auto_clear_workflow`
- `./scripts/ci/check_workflow_parsers`
- `./scripts/resolve-pr-threads.sh 202 --repo nathanjohnpayne/mergepath` → no unresolved threads

Note: `scripts/codex-review-check.sh 202` was failing before this approval because gate (b) had no reviewer-identity APPROVED review yet. This approval should satisfy that part; the remaining label-clear flow still depends on the normal Codex/current-head gate evaluation.

**Comment by nathanpayne-cursor** [comment]
@nathanjohnpayne — PR #202 is blocked only on the `needs-external-review` label.

You approved on HEAD 8a26f4e as nathanpayne-cursor, all 19 unit tests pass, 0 unresolved threads, all required CI checks green. Per [REVIEW_POLICY.md § Agent prohibitions](https://github.com/nathanjohnpayne/mergepath/blob/main/REVIEW_POLICY.md), agents do not remove this label.

When you're ready, clear it from any device:
- GitHub UI: Labels sidebar → click `x` on `needs-external-review`
- CLI: `gh pr edit 202 --remove-label needs-external-review`

I'll merge once the label is gone.

— posted by claude on behalf of `scripts/request-label-removal.sh`
